File: java/lld/state-machines/atm/ATMState.java
// The state interface as an abstract base: each action defaults to
// "not allowed here", so a concrete state overrides only its LEGAL actions.
public abstract class ATMState {
    protected final ATM atm;
    protected ATMState(ATM atm) { this.atm = atm; }

    public void insertCard()          { reject("insert card"); }
    public void enterPin(String pin)  { reject("enter PIN"); }
    public void withdraw(int amount)  { reject("withdraw"); }
    public void ejectCard()           { reject("eject card"); }

    public abstract String name();

    protected void reject(String action) {
        System.out.println("  x cannot " + action + " while " + name());
    }
}

File: typescript/lld/state-machines/atm/ATMState.ts
import { ATM } from './ATM';

// The state interface: every action the machine exposes. The abstract base
// gives each a default of "not allowed here", so a concrete state only
// overrides the actions that are LEGAL in its mode — the rest self-reject.
export abstract class ATMState {
    constructor(protected atm: ATM) {}

    insertCard(): void { this.reject('insert card'); }
    enterPin(_pin: string): void { this.reject('enter PIN'); }
    withdraw(_amount: number): void { this.reject('withdraw'); }
    ejectCard(): void { this.reject('eject card'); }

    abstract name(): string;

    protected reject(action: string): void {
        console.log(`  ✗ cannot ${action} while ${this.name()}`);
    }
}